### PR TITLE
UI for Scheduled Rule Changes

### PR DESCRIPTION
**Pull request by mozbhearsum**
This provides some basic UI for managing Scheduled Rule Changes. The drop down from "Rules" was the best I could come up with for Nav. I thought it made sense to keep it either there or as a link in the body of the page, as it's specifically related to Rules.

The UI for the new section is just a basic CRUD ui, largely a copy of the Rules stuff with extra fields added. I spent some time fussing with a datetime picker, because I couldn't think of any other reasonable way to configure that kind of change. You'll notice that it doesn't close automatically for some reason - I haven't been able to figure out why.

I added a button on the Rules page to create a Scheduled Change based on an existing Rule - I expect that's what we'll be using the vast majority of the time.

I made a couple of fixes related to error handling as part of this as well - there were some types of errors that resulted in no message in the UI (merely the console).

Running this patch along with https://github.com/mozilla/balrog/pull/103 should give you a fully working system that lets you add scheduled changes with the agent running to process them.

### CODE REVIEW

**Comment by nthomas-mozilla** [comment]
Some comments from playing around with the UI - it's a great start but I was able to poke some holes.

* creating a change a couple of minutes in the future coped with timezones nicely, yay
* when creating a change it wasn't obvious which timezone would be used until completing the selection
* the list of completed changes is nice, but leaves off the timezone on 'When'
* it's possible to schedule changes in the past without complaint, next agent run processes it. Often date pickers grey out the past, hopefully this is easy with our imports
* the UI around when vs telemetry should indicate it's either/or, maybe group them in a box
* 'Scheduled Change Details:' and 'Rule Details' kind of disappear compared to the bold text for the input boxes etc. Bold italics, or color, could emphasize them more
* on the 'Scheduled Rule Changes' page, using the 'Add a new ..' UI, what is the 'ID or Alias' textbox for ? Specifying an ID should be blocked, but Alias might be OK.
* error in the console - 'Please DO NOT specify the dropdownSelector option unless you are using jQuery AND Bootstrap.js. Please include jQuery AND Bootstrap.js, or write code to close the dropdown in the on-set-time callback. The dropdownSelector configuration option is being removed because it will not function properly.' at app.js:3:15106, when using 'Schedule a Change' on Rules page. Maybe a hint to make it close better
* not specifying a 'when' results in 'Error: b.when is undefined....' and the submission hangs (with or without Telemetry properties set)
* once we have many scheduled changes, it might be useful to have the 'Product/Channel filter' from the Rules page on Scheduled Rule Changes page. Not sure if that is hard^2, only nice to have at this point

**Comment by mozbhearsum** [comment]
Thanks for having such a good look at it! Most of those things should be pretty easy fixes. I saw that dropdownSelector error before, but I thought I'd fixed it. I'll look into it again.

We should probably check for timestamps being in the past on the backend as well as the frontend, I'll see about doing that.

**Comment by mozbhearsum** [comment]
On the "ID or Alias" box when creating a scheduled change via that page, I agree that we shouldn't encourage the scheduling of changes to existing rules there, but I don't think we can enforce it without backend changes.

At minimum, I think we should remove that entire field from that page - we can probably live without being able to set alias on a scheduled (arguably, that's undesirable).

If you think it's important to truly block it (in the sense that the backend will return 400), we'll probably want a separate endpoint or piece of request data to distinguish new scheduled changes that come in from the rules page vs. scheduled changes page.

**Comment by mozbhearsum** [comment]
Latest push addresses most of your comments, there's still a bit more to do though.

**Comment by mozbhearsum** [comment]
Alright, I think this should cover all of the key points you mentioned. I ended up hiding rule_id in all forms of the scheduled change modal because I thought it would be confusing to allow people to change the rule id when updating a scheduled change as well. That's another thing that we should probably reject in the backend, come to think of it.

**Comment by mozbhearsum** [comment]
Here's a compare link to see the changes since your last look: https://github.com/mozbhearsum/balrog-ui/compare/63672b98aa33f707ed428abfd6b7bf99ce0ba9df...86f31b5

**Comment by nthomas-mozilla** [comment]
This is looking really good. Tempted to r+ it because I think the comments that follow aren't huge issues, but would like to hear your response.

It's possible to schedule two changes on a rule for identical 'when', which leads to an attempt to merge them after applying one, and failure if they're incompatible (eg I changed the comment two different ways). They seem to be stuck then, failing to apply each time the agent runs. Bit of an edge case, but perhaps the backend should refuse changes if rule id and when match an existing change ?

Still an error if no When is specified for a change:
  Error: b.when is undefined
  b.addScheduledChange@http://localhost:8080/js/app.js:14:25739
  ...
Is that not expected to worked yet, so not a blocker ?  BTW, would like to chat to you about how you go about debugging that sort of issue when the js is uglified.

Some UX nice to haves:
- the date/time picker opens when clicking the calendar icon. Can we make the select box do the same thing, but only if it's empty ? Doubt we're going to type in a time in the right format from scratch, but editing might be useful.
- If your Scheduled Changes page is stale, and you try do delete an already applied change then the backend returns a 400 error, but the frontend keeps showing the 'Are you sure?' delete dialog.

**Comment by mozbhearsum** [comment]
Regarding two changes for the same Rule, I'm of two minds on it. It wasn't something I originally designed for, but since we don't yet have the needed hooks into Telemetry, I can see us wanting to schedule a sequence of throttling changes to the same rule. I agree that something needs to be done here - the main point of mergeUpdate and UpdateMergeError was to make sure that this sort of thing was dealt with when editing the Rule or Scheduled Change to avoid runtime errors like these.

I'm surprised that b.when error still exists, perhaps I glossed over it though. I'll see what I can do about the UX things as well. I had explicitly removed the open-calendar-on-input-click to enable editing of prepopulated times, but I agree that it's unintuitive for the case where that field is empty.

**Comment by mozbhearsum** [comment]
Still a bit more to fix up, but I improved the layout quite a bit, largely based on Njira's work from #32.

**Comment by mozbhearsum** [comment]
Alright, the latest version of this PR should address all your concerns other than the opening of the datetime picker when cliclking the input. I wasn't able to get that working without breaking something else.

Let me know when you want to chat about debugging, it would be good to chat about what to do about those sequential changes at the same time.

File: app/templates/rule_scheduled_change_modal.html [hunk 1]
```diff
@@ -0,0 +1,221 @@
+<div class="modal-header">
+    <h3 class="modal-title" ng-show="is_edit">Edit Scheduled Change</h3>
+    <h3 class="modal-title" ng-show="!is_edit">Add Scheduled Change</h3>
+</div>
+<div class="modal-body">
+  <form role="form">
+    <h4>Scheduled Change Details</h4>
+    <div class="row text-center">
+      <button id="btn_time" class="btn active" type="button" ng-click="toggleType('time')">Time Based</button>
+      <button id="btn_telemetry" class="btn" type="button" ng-click="toggleType('telemetry')">Telemetry Based</button>
```
Comment (nthomas-mozilla) [inline app/templates/rule_scheduled_change_modal.html]: I really like that having an either/or for When/Telemetry. Looks like we can add 'disabled' to class here so people don't try to use it yet.

**Comment by nthomas-mozilla** [comment]
Lets 🚢 it ... and follow up with some backend fixes for the multiple scheduled changes on a rule.

File: app/templates/rule_scheduled_change_modal.html (hunk 1 above)
Comment (mozbhearsum) [inline app/templates/rule_scheduled_change_modal.html]: Good idea. I'll do that before merging this.

**Comment by mozbhearsum** [comment]
I'm not sure why tc-gh didn't fire, but I created a Task by hand, which passed: https://tools.taskcluster.net/task-inspector/#O2NAdHPJTGeLFerkh7zpXg/0

**Pull request merged**